File: environment/src/main/java/jetbrains/exodus/tree/btree/BottomPage.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.tree.btree;

import jetbrains.exodus.ByteIterable;
import jetbrains.exodus.log.ByteIterableWithAddress;
import jetbrains.exodus.log.CompressedUnsignedLongByteIterable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.PrintStream;

class BottomPage extends BasePageImmutable {

    BottomPage(@NotNull final BTreeBase tree) {
        super(tree);
    }

    protected BottomPage(@NotNull final BTreeBase tree, @NotNull final ByteIterableWithAddress data) {
        super(tree, data);
    }

    protected BottomPage(@NotNull final BTreeBase tree, @NotNull final ByteIterableWithAddress data, int size) {
        super(tree, data, size);
    }

    @Override
    protected boolean isBottom() {
        return true;
    }

    @Override
    public long getChildAddress(int index) {
        return getKeyAddress(index);
    }

    @Override
    protected long getBottomPagesCount() {
        return 1;
    }

    @Override
    public ILeafNode get(@NotNull ByteIterable key) {
        return get(key, this);
    }

    @Override
    public ILeafNode find(@NotNull BTreeTraverser stack, int depth, @NotNull ByteIterable key, @Nullable ByteIterable value, boolean equalOrNext) {
        return find(stack, depth, key, value, equalOrNext, this);
    }

    @Override
    @NotNull
    protected BasePageMutable getMutableCopy(BTreeMutable treeMutable) {
        return new BottomPageMutable(treeMutable, this);
    }

    @Override
    public boolean keyExists(@NotNull ByteIterable key) {
        return keyExists(key, this);
    }

    @Override
    public boolean exists(@NotNull ByteIterable key, @NotNull ByteIterable value) {
        return exists(key, value, this);
    }

    @Override
    public boolean childExists(@NotNull ByteIterable key, long pageAddress) {
        return false;
    }

    @Override
    public String toString() {
        return "Bottom [" + size + "] @ " + (getDataAddress() - CompressedUnsignedLongByteIterable.getIterable(size << 1).getLength() - 1);
    }

    @Override
    public void dump(PrintStream out, int level, ToString renderer) {
        dump(out, level, renderer, this);
    }

    protected void reclaim(final ByteIterable keyIterable, @NotNull final BTreeReclaimTraverser context) {
        final BasePage node = context.currentNode;
        if (node.isBottom()) {
            if (node.getDataAddress() == getDataAddress()) {
                doReclaim(context);
                return;
            } else if (node.size > 0 && node.getMinKey().compareKeyTo(keyIterable) == 0) {
                // TODO: if root is empty just use other code in tree reclaim instead, omitting "node.size > 0" check
                // we are already in desired bottom page, but address in not the same
                return;
            }
        }
        // go up
        if (context.canMoveUp()) {
            while (true) {
                context.popAndMutate();
                context.moveRight();
                final int index = context.getNextSibling(keyIterable);
                if (index < 0) {
                    if (!context.canMoveUp()) {
                        context.moveTo(Math.max(-index - 2, 0));
                        break;
                    }
                } else {
                    context.pushChild(index); // node is always internal
                    break;
                }
            }
        }
        // go down
        while (context.canMoveDown()) {
            int index = context.getNextSibling(keyIterable);
            if (index < 0) {
                index = Math.max(-index - 2, 0);
            }
            context.pushChild(index);
        }
        if (context.currentNode.getDataAddress() == getDataAddress()) {
            doReclaim(context);
        }
    }

    static ILeafNode get(@NotNull ByteIterable key, @NotNull BasePage page) {
        final int index = page.binarySearch(key);
        if (index >= 0) {
            return page.getKey(index);
        }
        return null;
    }

    @NotNull
    private static ILeafNode findFirst(@NotNull BTreeTraverser stack, int depth, BasePage page) {
        final ILeafNode result;
        if (page.isBottom()) {
            result = page.getMinKey();
            stack.currentNode = page;
            stack.currentPos = 0;
            stack.top = depth;
        } else {
            result = findFirst(stack, depth + 1, page.getChild(0));
            stack.setAt(depth, new TreePos(page, 0));
        }
        return result;
    }

    @Nullable
    static ILeafNode find(@NotNull BTreeTraverser stack, int depth, @NotNull ByteIterable key, @Nullable ByteIterable value, boolean equalOrNext, @NotNull BasePage page) {
        int index = page.binarySearch(key);
        if (index < 0) {
            if (value == null && equalOrNext) {
                index = -index - 1;
                if (index >= page.getSize()) return null; // after last element - no element to return
            } else {
                return null;
            }
        }

        ILeafNode ln = page.getKey(index);

        if (ln.isDup()) {
            BasePage dupRoot = ln.getTree().getRoot();
            ILeafNode dupLeaf;
            if (value != null) {
                // move dup cursor to requested value
                dupLeaf = dupRoot.find(stack, depth + 1, value, null, equalOrNext);
                if (dupLeaf == null) {
                    return null;
                }
            } else {
                dupLeaf = findFirst(stack, depth + 1, dupRoot);
            }

            stack.setAt(depth, new TreePos(page, index));

            ((BTreeTraverserDup) stack).inDupTree = true;

            return dupLeaf;
        }
        if (stack.isDup()) {
            ((BTreeTraverserDup) stack).inDupTree = false;
        }


        if (value == null || (equalOrNext ? value.compareTo(ln.getValue()) <= 0 : value.compareTo(ln.getValue()) == 0)) {
            stack.currentNode = page;
            stack.currentPos = index;
            stack.top = depth;
            //TODO: stack.bottom = 0?
            return ln;
        }

        return null;
    }

    static boolean keyExists(@NotNull ByteIterable key, @NotNull BasePage page) {
        return page.binarySearch(key) >= 0;
    }

    static boolean exists(@NotNull ByteIterable key, @NotNull ByteIterable value, @NotNull BasePage page) {
        ILeafNode ln = page.get(key);
        return ln != null && ln.valueExists(value);
    }

    static void dump(PrintStream out, int level, ToString renderer, BasePage page) {
        indent(out, level);
        out.println(page);
        for (int i = 0; i < page.getSize(); i++) {
            page.getKey(i).dump(out, level + 1, renderer);
        }
    }

}

File: environment/src/main/kotlin/jetbrains/exodus/tree/btree/BottomPage.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.tree.btree

import jetbrains.exodus.*
import jetbrains.exodus.log.ByteIterableWithAddress
import jetbrains.exodus.log.CompressedUnsignedLongByteIterable.Companion.getIterable
import jetbrains.exodus.tree.Dumpable
import java.io.PrintStream
import kotlin.math.max

internal open class BottomPage : BasePageImmutable {
    constructor(tree: BTreeBase) : super(tree)
    constructor(
        tree: BTreeBase,
        data: ByteIterableWithAddress,
        loggableInsideSinglePage: Boolean
    ) : super(tree, data, loggableInsideSinglePage)

    constructor(
        tree: BTreeBase, data: ByteIterableWithAddress, size: Int,
        loggableInsideSinglePage: Boolean
    ) : super(tree, data, size, loggableInsideSinglePage)

    override val isBottom: Boolean
        get() = true

    override fun getChildAddress(index: Int): Long {
        return getKeyAddress(index)
    }

    override val bottomPagesCount: Long
        get() = 1

    override fun get(key: ByteIterable): ILeafNode? {
        return Companion[key, this]
    }

    override fun find(
        stack: BTreeTraverser,
        depth: Int,
        key: ByteIterable,
        value: ByteIterable?,
        equalOrNext: Boolean
    ): ILeafNode? {
        return find(stack, depth, key, value, equalOrNext, this)
    }

    override fun getMutableCopy(treeMutable: BTreeMutable): BasePageMutable {
        return BottomPageMutable(treeMutable, this)
    }

    override fun keyExists(key: ByteIterable): Boolean {
        return keyExists(key, this)
    }

    override fun exists(key: ByteIterable, value: ByteIterable): Boolean {
        return exists(key, value, this)
    }

    override fun childExists(key: ByteIterable, pageAddress: Long): Boolean {
        return false
    }

    override fun toString(): String {
        return "Bottom [" + size + "] @ " + (dataAddress - getIterable((size shl 1).toLong()).length - 1)
    }

    override fun dump(out: PrintStream, level: Int, renderer: Dumpable.ToString?) {
        dump(out, level, renderer, this)
    }

    fun reclaim(keyIterable: ByteIterable, context: BTreeReclaimTraverser) {
        val node = context.currentNode
        if (node.isBottom) {
            if (node.dataAddress == dataAddress) {
                doReclaim(context)
                return
            } else if (node.size > 0 && node.minKey.compareKeyTo(keyIterable) == 0) {
                // we are already in desired bottom page, but address in not the same
                return
            }
        }
        // go up
        if (context.canMoveUp()) {
            while (true) {
                context.popAndMutate()
                context.moveRight()
                val index = context.getNextSibling(keyIterable)
                if (index < 0) {
                    if (!context.canMoveUp()) {
                        context.moveTo(max(-index - 2, 0))
                        break
                    }
                } else {
                    context.pushChild(index) // node is always internal
                    break
                }
            }
        }
        // go down
        while (context.canMoveDown()) {
            var index = context.getNextSibling(keyIterable)
            if (index < 0) {
                index = max(-index - 2, 0)
            }
            context.pushChild(index)
        }
        if (context.currentNode.dataAddress == dataAddress) {
            doReclaim(context)
        }
    }

    companion object {
        operator fun get(key: ByteIterable, page: BasePage): ILeafNode? {
            val index = page.binarySearch(key)
            return if (index >= 0) {
                page.getKey(index)
            } else null
        }

        private fun findFirst(stack: BTreeTraverser, depth: Int, page: BasePage): ILeafNode {
            val result: ILeafNode
            if (page.isBottom) {
                result = page.minKey
                stack.currentNode = page
                stack.currentPos = 0
                stack.top = depth
            } else {
                result = findFirst(stack, depth + 1, page.getChild(0))
                stack.setAt(depth, TreePos(page, 0))
            }
            return result
        }

        fun find(
            stack: BTreeTraverser,
            depth: Int,
            key: ByteIterable,
            value: ByteIterable?,
            equalOrNext: Boolean,
            page: BasePage
        ): ILeafNode? {
            var index = page.binarySearch(key)
            if (index < 0) {
                if (value == null && equalOrNext) {
                    index = -index - 1
                    if (index >= page.size) return null // after last element - no element to return
                } else {
                    return null
                }
            }
            val ln: ILeafNode = page.getKey(index)
            if (ln.isDup) {
                val dupRoot = ln.tree.root
                val dupLeaf: ILeafNode?
                if (value != null) {
                    // move dup cursor to requested value
                    dupLeaf = dupRoot.find(stack, depth + 1, value, null, equalOrNext)
                    if (dupLeaf == null) {
                        return null
                    }
                } else {
                    dupLeaf = findFirst(stack, depth + 1, dupRoot)
                }
                stack.setAt(depth, TreePos(page, index))
                (stack as BTreeTraverserDup).inDupTree = true
                return dupLeaf
            }
            if (stack.isDup) {
                (stack as BTreeTraverserDup).inDupTree = false
            }
            if (value == null || (if (equalOrNext) value <= ln.value else value.compareTo(ln.value) == 0)) {
                stack.currentNode = page
                stack.currentPos = index
                stack.top = depth
                return ln
            }
            return null
        }

        fun keyExists(key: ByteIterable, page: BasePage): Boolean {
            return page.binarySearch(key) >= 0
        }

        fun exists(key: ByteIterable, value: ByteIterable, page: BasePage): Boolean {
            val ln = page[key]
            return ln != null && ln.valueExists(value)
        }

        fun dump(out: PrintStream, level: Int, renderer: Dumpable.ToString?, page: BasePage) {
            indent(out, level)
            out.println(page)
            for (i in 0 until page.size) {
                page.getKey(i).dump(out, level + 1, renderer)
            }
        }
    }
}
